Ensure the Minimum TLS version is consistently set

jentfoo opened the pull request:
We were already configuring the minimum TLS version for the server side in `utils/tls.go`. However this was not being consistently used, and client side TLS version requirements were not set.

With this PR I have introduced a minimum TLS version in `api/defaults/defaults.go`.  Currently 1.2, all server side listeners are designed to reference this value.  This should make future updates simpler. Client side TLS configurations are hard coded to 1.2, since 1.0 and 1.1 are deprecated.

This PR is not currently planned to be backported

zmb3 commented on api/defaults/defaults.go:
```diff
@@ -34,6 +35,9 @@ const (
 	// DefaultIdleTimeout is a default idle connection timeout.
 	DefaultIdleTimeout = 30 * time.Second
 
+	// MinTLSVersion is the minimum TLS version the server will accept.
```
The godoc here mentions only servers, but you're using this value when configuring clients.

In fact, I view this change as _weakening_ client side security. Go already defaults to minimum of 1.2 for TLS clients when `MinVersion` is not set. If at some point in the future, TLS 1.2 is no longer considered secure, Go may well change the default, and we would no longer be protected by the default behavior since we've explicitly opted in to TLS 1.2 everywhere.

jentfoo replied:
> but you're using this value when configuring clients

If so, that was unintentional.  I intended to reference this field for servers and for clients to reference `tls.VersionTLS12` directly.

> Go already defaults to minimum of 1.2 for TLS clients when MinVersion is not set

I did not realize this, I agree with you that we should remove this from client configurations.

Do you see anywhere I accidentally set `apidefaults.MinTLSVersion` in a client side configuration?  If not, I will just remove the `tls.` referenced changes.

zmb3 replied:
Ok, looking better, but now I wonder if the api/ module is the right place for this constant.

Our API module is designed to build clients, so why does it need a constant for how to configure servers?

jentfoo replied:
I put it here because it co-locates with `DefaultIOTimeout` and `DefaultIdleTimeout`.  These are also server configuration, but I believed they existed here to inform clients of this server configuration.  I think the TLS version configuration may make sense in the same way, particularly if / when we update to version 1.3.

Let me know your thoughts, I feel this is most consistent with the existing parameters.  But happy to move it.

zmb3 replied:
I'd like to see it moved. A constant in a module that is never used in said module feels like a code smell.

greedy52 commented on lib/srv/db/common/auth.go:
```diff
@@ -495,7 +495,9 @@ func (a *dbAuth) GetTLSConfig(ctx context.Context, sessionCtx *Session) (*tls.Co
 // Config also includes database specific adjustment.
 func (a *dbAuth) getTLSConfigVerifyFull(ctx context.Context, sessionCtx *Session) (*tls.Config, error) {
-	tlsConfig := &tls.Config{}
+	tlsConfig := &tls.Config{
+		MinVersion: tls.VersionTLS12,
+	}
```
`getTLSConfigVerifyFull` is used for Teleport Database Service to connect to actual database servers. I think this can potentially break if the database server cannot do TLS 1.2. Hopefully no such database servers exist in the field. We might want to update the documentation though as TLS 1.2 will be the minimum requirement or maybe a breaking change for v14.

zmb3 commented on lib/fixtures/keys.go:
```diff
@@ -113,7 +113,8 @@ func (t *TLSConfig) NewClient() *http.Client {
 			TLSClientConfig: &tls.Config{
-				RootCAs: t.CertPool,
+				RootCAs:    t.CertPool,
+				MinVersion: tls.VersionTLS12,
```
This is a client configuration.
```suggestion
```

zmb3 commented on lib/inventory/metadata/metadata.go:
```diff
@@ -106,7 +106,9 @@ func (c *fetchConfig) setDefaults() {
 			if transport.TLSClientConfig == nil {
-				transport.TLSClientConfig = &tls.Config{}
+				transport.TLSClientConfig = &tls.Config{
+					MinVersion: tls.VersionTLS12,
```
```suggestion
```

zmb3 commented on lib/srv/db/mysql/engine.go:
```diff
@@ -460,7 +460,9 @@ func FetchMySQLVersion(ctx context.Context, database types.Database) (string, er
 	if database.IsCloudSQL() {
-		dialer = newGCPTLSDialer(&tls.Config{})
+		dialer = newGCPTLSDialer(&tls.Config{
+			MinVersion: tls.VersionTLS12,
```
```suggestion
```

zmb3 commented on lib/srv/db/proxyserver.go:
```diff
@@ -626,6 +626,7 @@ func (s *ProxyServer) getConfigForServer(ctx context.Context, identity tlsca.Ide
 		ServerName:   server.GetHostname(),
 		Certificates: []tls.Certificate{cert},
 		RootCAs:      pool,
+		MinVersion:   tls.VersionTLS12,
```
```suggestion
```

zmb3 commented on lib/usagereporter/teleport/usagereporter.go:
```diff
@@ -165,6 +165,7 @@ func NewPrehogSubmitter(ctx context.Context, prehogEndpoint string, clientCert *
 		GetClientCertificate: func(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
 			return clientCert, nil
 		},
+		MinVersion: tls.VersionTLS12,
```
```suggestion
```

jentfoo committed:
Ensure server side Minimum TLS version is consistently set

We were already configuring the minimum TLS version for the server side in `utils/tls.go`.
With this commit I have introduced a minimum TLS version in `api/defaults/defaults.go`.  Currently 1.2, all server side listeners are designed to reference this value or use `utils/tls.go`.  This should make future updates simpler.

jentfoo commented:
Closing this PR and the pre-req PR's #30368 and #30193

Although the refactor seems ideal, it also seems like efforts here are too significant to justify this change.
